### PR TITLE
add docs on what privileges are needed for using metricbeat mongodb module

### LINKED ISSUE #13820: Document privileges needed to use Metricbeat MongoDB module
Check the privileges needed to use the MongoDB module and document them.

Add user and roles to MongoDB integration tests so these privileges are checked.

### PR DESCRIPTION
**Pull request by hwang381**
resolves elastic/beats#13820

### CODE REVIEW

**Comment by ChrsMark** [comment]
Hi @hwang381 thanks for working on this! In order to fix the failing test we need to execute `make update` under `beats/metricbeat` path.

File: metricbeat/module/mongodb/_meta/docs.asciidoc [hunk 1]
```diff
@@ -47,3 +47,27 @@ The default metricsets are `collstats`, `dbstats` and `status`.
...
+In order to use the metricsets, the MongoDB user specified in the module configuration needs to have certain https://docs.mongodb.com/manual/core/authorization/#privileges[privileges]
+
+Using different metricsets requires different set of privileges
+
+* `collstats` requires
+** https://docs.mongodb.com/manual/reference/privilege-actions/#top[`top` action] on https://docs.mongodb.com/manual/reference/resource-document/#cluster-resource[`cluster` resource]
+
+* `dbstats` requires
+** https://docs.mongodb.com/manual/reference/privilege-actions/#listDatabases[`listDatabases` action] on https://docs.mongodb.com/manual/reference/resource-document/#cluster-resource[`cluster` resource]
+** for each of the databases, also need https://docs.mongodb.com/manual/reference/privilege-action/#dbStats[`dbStats` action] on the https://docs.mongodb.com/manual/reference/resource-document/#database-and-or-collection-resource[`database` resource]
```
Comment (ChrsMark) [inline metricbeat/module/mongodb/_meta/docs.asciidoc]: dbStats seems to not working 🤔

File: metricbeat/module/mongodb/_meta/docs.asciidoc [hunk 2]
```diff
@@ -47,3 +47,27 @@ The default metricsets are `collstats`, `dbstats` and `status`.
...
+
+* `status` requires
+** https://docs.mongodb.com/manual/reference/privilege-actions/#serverStatus[`serverStatus` action] on https://docs.mongodb.com/manual/reference/resource-document/#cluster-resource[`cluster` resource]
+
+* `replstatus` requires
+** https://docs.mongodb.com/manual/reference/privilege-actions/#find[`find`/`listCollections` action] on https://docs.mongodb.com/manual/reference/local-database/[the `local` database resource]
+** https://docs.mongodb.com/manual/reference/privilege-actions/#collStats[`collStats` action] on https://docs.mongodb.com/manual/reference/local-database/#local.oplog.rs[the `local.oplog.rs` collection resource]
```
Comment (ChrsMark) [inline metricbeat/module/mongodb/_meta/docs.asciidoc]: What about the [`replSetGetStatus`](https://docs.mongodb.com/manual/reference/privilege-actions/#replSetGetStatus)?

Comment (hwang381) [reply metricbeat/module/mongodb/_meta/docs.asciidoc]: Ahh missed it. Will add

File: metricbeat/module/mongodb/_meta/docs.asciidoc (hunk 1 above)
Comment (hwang381) [reply metricbeat/module/mongodb/_meta/docs.asciidoc]: ooops my bad, will fix

**Comment by hwang381** [comment]
Thanks for the feedback @ChrsMark 

Regard documenting the privileges needed for every metricset in their respective doc page, do you think it would be better idea to **just** put the documentation in the respective metricset doc instead of putting it in the general module documentation?

File: metricbeat/module/mongodb/_meta/docs.asciidoc [hunk 3]
```diff
@@ -47,3 +47,28 @@ The default metricsets are `collstats`, `dbstats` and `status`.
 
 The MongoDB metricsets were tested with MongoDB 3.4 and 3.0 and are expected to
 work with all versions >= 2.8.
+
+[float]
+=== MongoDB Privileges
+
+In order to use the metricsets, the MongoDB user specified in the module configuration needs to have certain https://docs.mongodb.com/manual/core/authorization/#privileges[privileges]
```
Comment (jsoriano) [inline metricbeat/module/mongodb/_meta/docs.asciidoc]:
```suggestion
In order to use the metricsets, the MongoDB user specified in the module configuration needs to have certain https://docs.mongodb.com/manual/core/authorization/#privileges[privileges].
```

File: metricbeat/module/mongodb/_meta/docs.asciidoc [hunk 4]
```diff
@@ -47,3 +47,28 @@ The default metricsets are `collstats`, `dbstats` and `status`.
...
+
+Using different metricsets requires different set of privileges
```
Comment (jsoriano) [inline metricbeat/module/mongodb/_meta/docs.asciidoc]:
```suggestion
Using different metricsets requires different set of privileges:
```

File: metricbeat/module/mongodb/collstats/_meta/docs.asciidoc [hunk 5]
```diff
@@ -11,3 +11,7 @@ It is using the `top` administrative command to return usage statistics for each
 - update
 - remove
 - commands
+
+It requires the following privileges
```
Comment (jsoriano) [inline metricbeat/module/mongodb/collstats/_meta/docs.asciidoc]:
```suggestion
It requires the following privileges:
```
And similar in other metricsets.

**Comment by ChrsMark** [comment]
@hwang381 thanks for your work on this!

After revisiting the issue and mongodb docs your patch refers to, I would propose to add small snippets on each metricset's docs which will show how one can create users and grant the right privileges on them. For example on `collstats` I would add:

### Create user with proper privileges for the metricset

#### Create the role
```
db.createRole(
   {
     role: "collstatsRole",
     privileges: [
       { resource: { cluster: true }, actions: [ "top"] },
     ],
     roles: []
   }
)

```
#### create the privileged user  
```
db.createUser(
    {
        user: "beatstop",
        pwd: "pass",
        roles: [ "collstatsRole"]
    }
)
```
So, the users will be able to easily create the mongo-user required to run the spesific metricset, but also we would be able to confirm that the privileges we document will do the job.

Let me know what you think! @jsoriano feel free to add/comment on anything!

**Comment by jsoriano** [comment]
> After revisiting the issue and mongodb docs your patch refers to, I would propose to add small snippets on each metricset's docs which will show how one can create users and grant the right privileges on them.

+1 to that, maybe we can also add an example with a role with all the privileges in the module docs.

**Comment by hwang381** [comment]
I've integrated all the above suggestions :)

File: metricbeat/module/mongodb/dbstats/_meta/docs.asciidoc [hunk 6]
```diff
@@ -1 +1,44 @@
...
+In Mongo shell, you can use the following command to create the role
+
+....
+db.createRole(
+   {
+     role: "dbstatsRole",
+     privileges: [
+       { resource: { cluster: true }, actions: ["listDatabases"] },
+       // for each of the databases
+       { resource: { db: "db_1"}, actions: ["dbStats"]},
```
Comment (ChrsMark) [inline metricbeat/module/mongodb/dbstats/_meta/docs.asciidoc]: 🤔 this one is not working for me giving: `E  QUERY    [js] uncaught exception: Error: resource must set both db and collection or neither, but not exactly one. :`

Had to change it to:
```
db.createRole(
   {
     role: "dbstatsRole",
     privileges: [
       { resource: { cluster: true }, actions: ["listDatabases"] },
       { resource: { db: "testdb", collection: ""}, actions: ["dbStats"]}
     ],
     roles: []
   }
)
```

File: metricbeat/docs/modules/mongodb.asciidoc [hunk 7]
```diff
@@ -55,6 +55,72 @@ The default metricsets are `collstats`, `dbstats` and `status`.
...
+** https://docs.mongodb.com/manual/reference/privilege-actions/#find[`find`/`listCollections` action] on https://docs.mongodb.com/manual/reference/local-database/[the `local` database resource]
+** https://docs.mongodb.com/manual/reference/privilege-actions/#collStats[`collStats` action] on https://docs.mongodb.com/manual/reference/local-database/#local.oplog.rs[the `local.oplog.rs` collection resource]
+** https://docs.mongodb.com/manual/reference/privilege-actions/#replSetGetStatus[`replSetGetStatus` action] on https://docs.mongodb.com/manual/reference/resource-document/#cluster-resource[`cluster` resource]
+
+* `metrics` requires
+** https://docs.mongodb.com/manual/reference/privilege-actions/#serverStatus[`serverStatus` action] on https://docs.mongodb.com/manual/reference/resource-document/#cluster-resource[`cluster` resource]
+
+In Mongo shell, you can use the following command to create the role that encompass all of the privileges
+
+....
```
Comment (jsoriano) [inline metricbeat/docs/modules/mongodb.asciidoc]: I think that code should be enclosed in source tags like this:
```suggestion
["source","js",subs="attributes"]
----
```
And same thing in other code snippets.

File: metricbeat/docs/modules/mongodb.asciidoc [hunk 8]
```diff
@@ -55,6 +55,72 @@ The default metricsets are `collstats`, `dbstats` and `status`.
...
+       { resource: { cluster: true }, actions: ["top", "listDatabases", "serverStatus", "replSetGetStatus"] },
+       // for each of the databases you want to use dbstats metricset with
+       { resource: { db: "db_1"}, actions: ["dbStats"]},
+       { resource: { db: "db_2"}, actions: ["dbStats"]},
+
+     ],
+     roles: []
+   }
+)
+....
```
Comment (jsoriano) [inline metricbeat/docs/modules/mongodb.asciidoc]: Also when closing the block, use hyphens.
```suggestion
----
```

File: metricbeat/module/mongodb/_meta/docs.asciidoc [hunk 9]
```diff
@@ -47,3 +47,69 @@ The default metricsets are `collstats`, `dbstats` and `status`.
...
+
+* `replstatus` requires
+** https://docs.mongodb.com/manual/reference/privilege-actions/#find[`find`/`listCollections` action] on https://docs.mongodb.com/manual/reference/local-database/[the `local` database resource]
+** https://docs.mongodb.com/manual/reference/privilege-actions/#collStats[`collStats` action] on https://docs.mongodb.com/manual/reference/local-database/#local.oplog.rs[the `local.oplog.rs` collection resource]
+** https://docs.mongodb.com/manual/reference/privilege-actions/#replSetGetStatus[`replSetGetStatus` action] on https://docs.mongodb.com/manual/reference/resource-document/#cluster-resource[`cluster` resource]
+
+* `metrics` requires
+** https://docs.mongodb.com/manual/reference/privilege-actions/#serverStatus[`serverStatus` action] on https://docs.mongodb.com/manual/reference/resource-document/#cluster-resource[`cluster` resource]
+
+In Mongo shell, you can use the following command to create the role that encompass all of the privileges
```
Comment (jsoriano) [inline metricbeat/module/mongodb/_meta/docs.asciidoc]:
```suggestion
In Mongo shell, you can use the following command to create the role that encompass all of the privileges.
```

File: metricbeat/module/mongodb/_meta/docs.asciidoc [hunk 10]
```diff
@@ -47,3 +47,69 @@ The default metricsets are `collstats`, `dbstats` and `status`.
...
+       { resource: { db: "db_1"}, actions: ["dbStats"]},
+       { resource: { db: "db_2"}, actions: ["dbStats"]},
+
+     ],
+     roles: []
+   }
+)
+....
+
+You can then use the following command to create the privileged user
```
Comment (jsoriano) [inline metricbeat/module/mongodb/_meta/docs.asciidoc]:
```suggestion
You can then use the following command to create the privileged user.
```

And please review similar lines in other files.

File: metricbeat/module/mongodb/_meta/docs.asciidoc [hunk 11]
```diff
@@ -47,3 +47,69 @@ The default metricsets are `collstats`, `dbstats` and `status`.
...
+db.createUser(
+    {
+        user: "beats",
+        pwd: "pass",
+        roles: ["beatsRole"]
+    }
+)
+....
+
+or grant the role to an existing user
```
Comment (jsoriano) [inline metricbeat/module/mongodb/_meta/docs.asciidoc]:
```suggestion
To grant the role to an existing user, you can use the following command.
```
And please review similar lines in other files.

**Comment by hwang381** [comment]
@chrsblck Sure thing (I should've probably done that before I push those changes anyway, my bad :( )
Do you just start a MongoDB server locally and test those instructions, or are there something metricbeats specific to be done?

**Comment by ChrsMark** [comment]
In order to verify that the instructions are syntactically valid you just need a local mongodb (I run it in a Docker container `docker run -d -p 27017-27019:27017-27019 --name mongodb mongo`) to execute the commands on the mongo shell.

Then in order to verify that those commands have the proper impact you should run mongodb Metricbeat module and verify that each metricset has actually the right privileges required.

**Comment by ChrsMark** [comment]
> @chrsblck Sure thing (I should've probably done that before I push those changes anyway, my bad :( )
> Do you just start a MongoDB server locally and test those instructions, or are there something metricbeats specific to be done?

In order to verify that the instructions are syntactically valid you just need a local mongodb (I run it in a Docker container `docker run -d -p 27017-27019:27017-27019 --name mongodb mongo`) to execute the commands on the mongo shell.

**Comment by jsoriano** [comment]
We have found that it is common to just grant the [`clusterMonitor` role](https://docs.mongodb.com/manual/reference/built-in-roles/#clusterMonitor) to monitoring users. I have checked and it includes all the permissions you found that are needed for the Metricbeat module. It seems safe to use this role as it is maintained specifically for this use case.

For simplicity I think that we could provide instructions for using only this role in the module documentation. We could still list the permissions required in each metricset documentation.

@hwang381 sorry for the change of direction, but what would you think about doing this change?

**Comment by hwang381** [comment]
@jsoriano Great find. Now I am wondering if we should just mention the use of this one privilege only in the overall doc. If this privilege is purposely built for monitoring needs, then I wouldn't hestitate to just use this one.

**Comment by jsoriano** [comment]
> @jsoriano Great find. Now I am wondering if we should just mention the use of this one privilege only in the overall doc. If this privilege is purposely built for monitoring needs, then I wouldn't hestitate to just use this one.

Yeah, we should probably go towards using just this one. Could you make the changes in this PR to use only this permission? Thanks!

**Comment by hwang381** [comment]
Made some updates
@jsoriano The module doc page has been updated to use `clusterMonitor` role; several misc. styling issues have also been addressed
@ChrsMark All mongo shell commands added should at least run successfully now. I am still not too sure about whether every command is truly least privilege though. Probably need to dive deeper.

**Comment by ChrsMark** [comment]
@hwang381 thank's for the update!

I would say that we could summarise this as follows:
1. List in Module docs all the required privileges.
2. Mention that `clusterMonitor` role is required for this module and provide the command and reference link on how to create and grant this role.
3. On each Metricset's docs just mention the specific privilege that is required and that is covered by `clusterMonitor` role.

In this I think we don't need to verify that "every command is truly least privilege" because the usage of `clusterMonitor` is the recommended way to go.

File: metricbeat/module/mongodb/_meta/docs.asciidoc [hunk 12]
```diff
@@ -47,3 +47,30 @@ The default metricsets are `collstats`, `dbstats` and `status`.
...
+    }
+)
+----
+
+You can use the following command in Mongo shell to grant the role to an existing user (make sure you are using the `admin` db by using `db` command in Mongo shell).
+
+["source","js",subs="attributes"]
+----
+db.grantRolesToUser("user", ["clusterMonitor"])
+----
```
Comment (jsoriano) [inline metricbeat/module/mongodb/_meta/docs.asciidoc]: Great :slightly_smiling_face:

File: metricbeat/module/mongodb/dbstats/_meta/docs.asciidoc [hunk 13]
```diff
@@ -1 +1,44 @@
...
+     privileges: [
+       { resource: { cluster: true }, actions: ["listDatabases"] },
+       // for each of the databases
+       { resource: { db: "db_1", collection: ""}, actions: ["dbStats"]},
+       { resource: { db: "db_2", collection: ""}, actions: ["dbStats"]}
+     ],
+     roles: []
+   }
+)
+----
```
Comment (jsoriano) [inline metricbeat/module/mongodb/dbstats/_meta/docs.asciidoc]: As @ChrsMark mentions, I think we can remove the example commands from the metricsets, and just list the required privileges for reference, mentioning that they are included in the `clusterMonitor` role.

**Comment by hwang381** [comment]
Removed command example in specific metricset docs and mentioned that the privileges are covered by clusterMonitor role

**Comment by jsoriano** [comment]
ok to test

**Comment by ChrsMark** [comment]
jenkins, test this again please

**Comment by ChrsMark** [comment]
Failing tests are irrelevant but we could wait until https://github.com/elastic/beats/pull/14179 is merged which might make CI happy. wdyt @jsoriano?

**Comment by jsoriano** [comment]
@ChrsMark I think we can merge this, metricbeat builds are green.

**Comment by ChrsMark** [comment]
I will go along! Merging this!

Thank you @hwang381 for your work on this!

**Comment by ChrsMark** [comment]
I will go along and merge this!

Thank you @hwang381 for working on this!

**Pull request merged**